embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/compression/CompressionOutputStream.kt

```diff
@@ -0,0 +1,70 @@
+package io.embrace.android.embracesdk.internal.compression
+
+import java.io.OutputStream
+import java.util.zip.GZIPOutputStream
+
+/**
+ * This Output Stream is used to compress the data that is being written to it.
+ * It will only compress the data if it's not already compressed.
+ * Given we started compressing payloads to disk from version 6.3.0 onwards,
+ * we need to check if the incoming data is already compressed or not for backwards compatibility.
+ */
+internal class CompressionOutputStream(
+    private val outputStream: OutputStream,
+) : OutputStream() {
+
+    private val magicNumberBuffer = mutableListOf<Byte>()
+    private var drainedBuffer: Boolean = false
+    private lateinit var impl: OutputStream
+
+    override fun write(b: Int) {
+        if (magicNumberBuffer.size < 2) {
+            magicNumberBuffer.add(b.toByte())
+        } else {
+            drainBufferIfNeeded()
+            impl.write(b)
+        }
+    }
+
+    override fun close() {
+        drainBufferIfNeeded()
+        impl.close()
+    }
+
+    /**
+     * If the buffer is not drained, it means that we have not yet written the first two bytes
+     * of the file, which are the gzip magic number. We need to write them before writing the rest
+     * of the file.
+     */
+    private fun drainBufferIfNeeded() {
+        if (!drainedBuffer) {
+            impl = generateOutputStream()
+            magicNumberBuffer.forEach {
+                impl.write(it.toInt())
+            }
+            drainedBuffer = true
+        }
+    }
+
+    private fun generateOutputStream(): OutputStream {
```

Comment:
If 0 or 1 byte is written, this will always create a GZipOutputStream. This can happen if `close()` is called too before 2 or more bytes are written. 

Does it make sense to modify the code so the stream is only set if two bytes are buffered, and if the drain happens before that, you just use the stream passed in?

One way you can do that is separating the stream creation method from the drain method, and only call the former in the `write()` method, and have the latter defer to stream passed in if the buffered stream isn't set. There are probably other ways to do this so feel free to go some other route.

Reply:
I made a modification to return the passed stream in the scenario when the buffer is empty because no bytes were written. But for the case were, the output stream is closed and one byte was written, we still want to compress that byte. That's why for that case, we still use the GzipOutputStream.

Reply:
Got it. Thanks!